scripts/ember_customization.py

```diff
@@ -161,3 +161,82 @@ def apply_custom_pf_constraint(n,
+        power_buses = n.buses.query("carrier == 'AC' and country in @unique_countries")[['x', 'y', 'country']]
+        power_buses = power_buses.reset_index().rename(
+                  columns={
+                              'Bus': 'bus_id', 
+                              'x': 'bus_x',    
+                              'y': 'bus_y'      
+                          }
+                                                       )
+        if power_buses.empty:
+            continue
+        sub_df = sub_df.reset_index(drop=True)
+        sub_df['plant_id'] = sub_df.index
+        pairs = pd.merge(sub_df, power_buses, on='country')
+        pairs['dx'] = pairs['x'] - pairs['bus_x']
+        pairs['dy'] = pairs['y'] - pairs['bus_y']
+        pairs['dist'] = (pairs['dx']**2 + pairs['dy']**2)**0.5
+        min_dist_idx = pairs.groupby('plant_id')['dist'].idxmin()
+        nearest_pairs = pairs.loc[min_dist_idx]
+        nearest_pairs['nearest_bus'] = nearest_pairs['bus_id']
+        nearest_pairs['heat_bus'] = nearest_pairs['nearest_bus'] + ' urban central heat'
+        valid_buses = n.buses.index.tolist()
+        nearest_pairs = nearest_pairs.query('heat_bus in @valid_buses')
+        
+        if nearest_pairs.empty:
+            continue
+        nearest_pairs['eff'] = nearest_pairs['efficiency'].fillna(0.45)
+        link_names = (nearest_pairs['nearest_bus'] + '_' + map_carrier + '_chp_' + nearest_pairs['name'].str.replace(' ', '_')).tolist()
+        bus0s = [f"EU {map_carrier}"] * len(nearest_pairs)
+        bus1s = nearest_pairs['nearest_bus'].tolist()
+        bus2s = nearest_pairs['heat_bus'].tolist()
+        bus3s = ["co2 atmosphere"] * len(nearest_pairs)
+        carriers_list = [f"urban central {map_carrier} CHP"] * len(nearest_pairs)
+        p_noms = (nearest_pairs['p_nom'] / nearest_pairs['eff']).tolist()
+        efficiencies = nearest_pairs['eff'].tolist()
+        efficiency2s = [0.40] * len(nearest_pairs)
+        efficiency3s = [costs.at[map_carrier, 'CO2 intensity']] * len(nearest_pairs)
+        marginal_costs = [costs.at[map_carrier, 'VOM']] * len(nearest_pairs)
+        capital_costs = [0] * len(nearest_pairs)
+        lifetimes = [25] * len(nearest_pairs)
+        p_nom_extendables = [False] * len(nearest_pairs)
+        reverseds = [False] * len(nearest_pairs)
+        
+        if link_names:
+            n.add(
+                "Link",
+                link_names,
+                bus0=bus0s,
+                bus1=bus1s,
+                bus2=bus2s,
+                bus3=bus3s,
+                carrier=carriers_list,
+                p_nom_extendable=p_nom_extendables,
+                p_nom=p_noms,
+                capital_cost=capital_costs,
+                marginal_cost=marginal_costs,
+                efficiency=efficiencies,
+                efficiency2=efficiency2s,
+                efficiency3=efficiency3s,
+                lifetime=lifetimes,
+                reversed=reverseds
```

Comment:
```suggestion
                bus0=f"EU {map_carrier}",
                bus1=nearest_pairs['nearest_bus'].tolist(),
                bus2=nearest_pairs['heat_bus'].tolist(),
                bus3="co2 atmosphere",
                carrier=f"urban central {map_carrier} CHP",
                p_nom_extendable=False,
                p_nom=(nearest_pairs['p_nom'] / nearest_pairs['eff']).tolist(),
                capital_cost=0,
                marginal_cost=costs.at[map_carrier, 'VOM'],
                efficiency=nearest_pairs['eff'].tolist(),
                efficiency2=0.4,
                efficiency3=costs.at[map_carrier, 'CO2 intensity'],
                lifetime=25,
                reversed=False
```

- can the lifetime be taken from costs, e.g. `costs.at[map_carrier, 'lifetime']` ? I guess it doesn't exist for coal/lignite CHPs?
- to we need .tolist() ?
- can you check if arguments passed as non-list (e.g. `bus0=f"EU {map_carrier}"` works? otherwise use your code with multiplying by len() ...

Reply:
Passing non-list arguments like bus0=f"EU {map_carrier}" to n.add( ) fails because pypsa requires lists matching the length of link_names for batch additions, causing a mismatch error( e.g, len(link_names)=len(bus0)=len(bus1)). In your suggestion, len(p_nom_extendable) is not equal to len(bus0).